aptos-move/aptos-vm/src/move_vm_ext/session.rs

```diff
@@ -211,49 +214,38 @@ impl<'r, 'l> SessionExt<'r, 'l> {
                         .entry(resource_group)
                         .or_insert_with(AccountChangeSet::new)
                         .add_resource_op(struct_tag, blob_op)
-                        .map_err(|_| common_error.clone())?;
+                        .map_err(|_| common_error())?;
                 } else {
-                    change_set_filtered
-                        .add_resource_op(addr, struct_tag, blob_op)
-                        .map_err(|_| common_error.clone())?;
+                    resources_filtered.insert(struct_tag, blob_op);
                 }
             }
 
-            for (name, blob_op) in modules {
-                change_set_filtered
-                    .add_module_op(ModuleId::new(addr, name), blob_op)
-                    .map_err(|_| common_error.clone())?;
-            }
+            change_set_filtered
+                .add_account_changeset(
+                    addr,
+                    AccountChangeSet::from_modules_resources(modules, resources_filtered),
+                )
+                .map_err(|_| common_error())?;
 
             for (resource_tag, resources) in resource_groups {
-                let source_data = remote
-                    .get_resource_group_data(&addr, &resource_tag)
-                    .map_err(|_| common_error.clone())?;
-                let (mut source_data, create) = if let Some(source_data) = source_data {
-                    let source_data =
-                        bcs::from_bytes(&source_data).map_err(|_| common_error.clone())?;
-                    (source_data, false)
-                } else {
-                    (BTreeMap::new(), true)
-                };
+                let mut source_data = remote
+                    .release_resource_group_cache(&addr, &resource_tag)
+                    .unwrap_or_default();
+                let create = source_data.is_empty();
```

Comment:
let's add a comment here to preserve why we do this for future generations :-) i.e. the idempotency